vsphere_plugin_common/__init__.py

```diff
@@ -766,43 +850,133 @@ def get_dvswitches(self):
+                        vswitch_name=vswitch_name,
+                        host_name=host.name,
+                    )
+                )
+                try:
+                    network_system.AddPortGroup(specification)
+                except vim.fault.AlreadyExists:
+                    # We tried to create it on a previous pass, but didn't see
+                    # any confirmation (e.g. due to a problem communicating
+                    # with the vCenter)
+                    # However, we shouldn't have reached this point if it
+                    # existed before we tried to create it anywhere, so it
+                    # should be safe to proceed.
+                    pass
+                runtime_properties['created_on'].append(host.name)
+
+            if self.port_group_is_on_all_hosts(port_group_name):
+                runtime_properties['status'] = 'created'
             else:
-                raise cfy_exc.NonRecoverableError(
-                    '{vswitch} was not a valid vswitch name. The valid '
-                    'vswitches are: {vswitches}'.format(
-                        vswitch=vswitch_name,
-                        vswitches=', '.join(vswitches),
+                return ctx.operation.retry(
+                    'Waiting for port group {name} to be created on all '
+                    'hosts.'.format(
+                        name=port_group_name,
                     )
                 )
 
-        for host in self.get_host_list():
-            network_system = host.configManager.networkSystem
-            specification = vim.host.PortGroup.Specification()
-            specification.name = port_group_name
-            specification.vlanId = vlan_id
-            specification.vswitchName = vswitch_name
-            vswitch = network_system.networkConfig.vswitch[0]
-            specification.policy = vswitch.spec.policy
-            ctx.logger.debug(
-                'Adding port group {group_name} to vSwitch {vswitch_name} on '
-                'host {host_name}'.format(
-                    group_name=port_group_name,
-                    vswitch_name=vswitch_name,
-                    host_name=host.name,
-                )
+    def port_group_is_on_all_hosts(self, port_group_name, distributed=False):
+        hosts, port_groups = self._get_port_group_host_count(
+            port_group_name,
+            distributed,
+        )
+        return hosts == port_groups
+
+    def port_group_is_on_any_hosts(self, port_group_name, distributed=False):
+        _, port_groups = self._get_port_group_host_count(
+            port_group_name,
+            distributed,
+        )
+        return port_groups > 0
+
+    def _get_port_group_host_count(self, port_group_name, distributed=False):
+        hosts = self.get_host_list()
+        host_count = len(hosts)
+
+        port_groups = self.get_obj_list([vim.Network])
+
+        if distributed:
+            port_groups = [
```

Comment:
Updated note: No, that doesn't work quite so well as planned, because one has to check if isinstance(thing) and the other if it's not an instance of that thing.

I'm not convinced there's a clean way to do this after playing with a couple of alternatives.